Proto reflection API

ericgribkoff opened the pull request:
protobuf: Proto reflection service

This includes the core and compiler changes from [https://github.com/grpc/grpc-java/pull/2380](https://github.com/grpc/grpc-java/pull/2380)

ericgribkoff commented:
@carl-mastrangelo @ejona86  - fyi, I updated this to match the merged PR for the compiler changes [https://github.com/grpc/grpc-java/pull/2380](https://github.com/grpc/grpc-java/pull/2380)

ejona86 commented on protobuf/build.gradle:
```diff
@@ -15,6 +26,8 @@ dependencies {
     }
 }
 
+configureProtoCompilation()
```
I think we've been putting services in the `services/` project (and the `io.grpc.protobuf.service` package).

ejona86 commented on protobuf/src/main/java/io/grpc/protobuf/ProtoReflectableServerBuilder.java:
```diff
@@ -0,0 +1,163 @@
+ * <p>This builder is designed to wrap another {@link ServerBuilder} and delegate to it for the
+ * actual server creation.
+ */
+public final class ProtoReflectableServerBuilder
```
I'm uneasy about wrapping the server builder, since each server builder as their own special bits. What did you learn about how C++ handles learning the available services?

ericgribkoff replied:
C++ reflection grabs a list of all service names registered at server start up. 

This wrapper still allows using all of the special bits of specific ServerBuilders, e.g.,

```
ProtoReflectableServerBuilder.forBuilder(
     NettyServerBuilder.forPort(port)
     .customNettyStuff()
).addService(..)
```

But maybe there's a better way to accomplish this

ericgribkoff commented:
I updated the PR. The reflectable server builder is gone, replaced with a new interface NotifyOnServerBuild that the reflection service implements. The interface specifies a method for the server builder to tell the service what server it is a part of when build() is called. Since the reflection service is now given a reference to the server, there is also a new method on Server (and the internal and fallback registries) to retrieve a list of the available services. I believe I have labeled everything with `@ExperimentalAPI`.

Is it safe to assume that service names are unique, even when combining the InternalHandlerRegistry and a fallback (mutable) registry? If this isn't true, the reflection responses may not contain all relevant information. The proto reflection services also uses service names to monitor whether services have been added or remove. If a service changes but keeps the same name, the reflection service wouldn't track chose changes. If this is an issue, it might be better for the reflection service to use the underlying protobuf descriptors instead of names.

carl-mastrangelo commented on core/src/main/java/io/grpc/HandlerRegistry.java:
```diff
@@ -42,6 +43,12 @@
+   * Returns the {@link ServerServiceDefinition}s provided by the registry.
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+  public abstract Collection<ServerServiceDefinition> getServices();
```
Any chance this could be a List or a Set?  Something that more strongly conveys what the type a user will get back?

It makes it easier to test (since you dont have to worry about order).

carl-mastrangelo replied:
FYI this will break existing concrete implementations since they won't have this method.

Are usual work around has been to implement the method, but make it throw UnsupportedOperationException

carl-mastrangelo replied:
Is the returned structure mutable?

carl-mastrangelo commented on core/src/main/java/io/grpc/NotifyOnServerBuild.java:
```diff
@@ -0,0 +1,43 @@
+ * the interface after build() has been called and before the {@link Server} instance is returned.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+public interface NotifyOnServerBuild {
```
I would very much like it if this was named "InternalNotifyOnServerBuild" or something to that effect.  Something that shows this is not public.  

Users will typically use tab completion and skip reading the docs and annotations.  Since the project is so young at this point, it would be wise to not expand the API in ways that hurt changing it later.

carl-mastrangelo commented on core/src/main/java/io/grpc/Server.java:
```diff
@@ -64,6 +65,12 @@ public int getPort() {
+   * Returns the services registered with the server.
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+  public abstract Collection<ServerServiceDefinition> getServices();
```
Same comments about API as above.

carl-mastrangelo commented on core/src/main/java/io/grpc/ServerBuilder.java:
```diff
@@ -86,6 +86,17 @@
+   *     interfaces
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+  public abstract <S extends NotifyOnServerBuild & BindableService> T addService(
```
Meta comment: This is the first time I have ever seen dual interface in generics actually used in all my java experience.

kudos.

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/AbstractServerImplBuilder.java:
```diff
@@ -78,6 +85,9 @@
   private final ArrayList<ServerTransportFilter> transportFilters =
       new ArrayList<ServerTransportFilter>();
 
+  private final ArrayList<NotifyOnServerBuild> notifyOnBuildList =
```
s/ArrayList/List/

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/ServerImpl.java:
```diff
@@ -176,6 +178,15 @@ public int getPort() {
+  @Override
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+  public Collection<ServerServiceDefinition> getServices() {
+    ArrayList<ServerServiceDefinition> services =
```
List

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,326 @@
+/**
+ * Provides a reflection service for Protobuf services (including the reflection service itself).
+ */
+public final class ProtoReflectionService extends ServerReflectionGrpc.ServerReflectionImplBase
```
Mark this experimental?

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,326 @@
+   * Retrieves the services registered to the server at build time, and stores those services with a
+   * {@link ProtoFileDescriptorSupplier}.
+   */
+  public void notifyOnBuild(Server server) {
```
Should this be an override?

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,326 @@
+   * {@link ProtoFileDescriptorSupplier}.
+   */
+  public void notifyOnBuild(Server server) {
+    this.server = server;
```
You should probably check this with `this.server = checkNotNull(server, "server")`

Also, it might be good to check that a server hasn't already been set, in case it is errantly called from some other place.

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,326 @@
+  private void processService(ServiceDescriptor service, FileDescriptor fd) {
+    fileDescriptorsBySymbol.put(service.getFullName(), fd);
+    for (MethodDescriptor method : service.getMethods()) {
+      fileDescriptorsBySymbol.put(method.getFullName(), fd);
```
this should check if there is already one by the same name.

ericgribkoff replied:
Even though it's using the method's fully-qualified name? From David's comments on the reflection API design doc, I thought that duplicates were impossible.

carl-mastrangelo replied:
I would not make that assumption.  There is very little enforcement of that, if at all.  David's comments are borne from C++, which treats all protos as part of the binary and thus may do such deduplication.  Java does not.

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,326 @@
+  private void processExtension(FieldDescriptor extension, FileDescriptor fd) {
+    String extensionName = extension.getContainingType().getFullName();
+    int extensionNumber = extension.getNumber();
+    if (fileDescriptorsByExtensionAndNumber.containsKey(extensionName)) {
```
can this if check be reversed?  Basically:

```
if (absent) {
  put empty map
}
insert entry
```


That reduce indentation and avoids copying code.

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,326 @@
+   * Processes the service file descriptors. This prepares a list of the service names and indexes
+   * the file descriptors by name, symbol, and extension.
+   */
+  private synchronized void initFileDescriptorMaps() {
```
why synchronized?

ericgribkoff replied:
I wanted to lazily load the protobuf file descriptors, while avoiding processing them for each incoming request. The synchronized is to stop multiple StreamObservers from simultaneously accessing the lookup maps during initialization.

carl-mastrangelo replied:
I guess my question is more, why is this one method synchronized, unlike the rest of the methods in this class.

carl-mastrangelo replied:
Ok, but this is getting into the realm of tricky.  This method is synchronized, but the methods that access the various maps here are not.  For example, Your stream observer could see memory that is being modified by this methods.  That would be a race.

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,326 @@
+    for (ServerServiceDefinition service : services) {
+      io.grpc.ServiceDescriptor serviceDescriptor = service.getServiceDescriptor();
+      if (serviceDescriptor.getMarshallerDescriptor() != null
+          && serviceDescriptor.getMarshallerDescriptor()
```
break lines at the highest semantic level.  In thi case the instance of should be on the same line as both its args.

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,326 @@
+        switch (request.getMessageRequestCase()) {
+          case FILE_BY_FILENAME:
+            getFileByName(request);
+            break;
```
return?

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,326 @@
+            listServices(request);
+            break;
+          default:
+            sendErrorResponse(request, Status.UNIMPLEMENTED, "");
```
What does empty string mean here?

carl-mastrangelo commented:
To answer your specific question about service names being unique: no, do not assume anything.  The data you are getting comes from descriptor protos which don't have strong checking against this.  You can explicitly disallow it, but you can't assume it won't happen.

carl-mastrangelo commented on services/src/main/proto/io/grpc/reflection/v1alpha/reflection.proto:
```diff
@@ -0,0 +1,155 @@
+  // This field uses the error codes defined in grpc::StatusCode.
+  int32 error_code = 1;
+  string error_message = 2;
+}
```
add a newline?

carl-mastrangelo commented on services/src/test/java/io/grpc/protobuf/service/ProtoReflectionServiceTest.java:
```diff
@@ -0,0 +1,312 @@
+
+  @Test
+  public void listServices() throws Exception {
+    HashSet<ServiceResponse> originalServices = new HashSet<ServiceResponse>(
```
Set<>

carl-mastrangelo commented on services/src/test/java/io/grpc/protobuf/service/ProtoReflectionServiceTest.java:
```diff
@@ -0,0 +1,312 @@
+
+  @Test
+  public void fileByFilename() throws Exception {
+    final ServerReflectionRequest request =
```
Don't use `final` for local variables.  It doesn't usually buy you much, and it adds boiler plate.

carl-mastrangelo commented on services/src/test/java/io/grpc/protobuf/service/ProtoReflectionServiceTest.java:
```diff
@@ -0,0 +1,312 @@
+        reflectionService.serverReflectionInfo(responseObserver);
+    requestObserver.onNext(request);
+    requestObserver.onCompleted();
+    assertEquals(goldenResponse, responseObserver.firstValue().get());
```
Make sure to split tests into 3 distinct sections: Setup, test, verify.  Each section should be separated by a newline to make it visually easier to parse.

(It also has a nice side effect that if it is hard to fit your test into this structure, it means the test is testing too much.  You're code is fine here though)

ericgribkoff commented:
I added a commit that addresses most of the comments. 

For the duplicate service names issue, I "resolved" the issue by taking the first service for each name from the return value of getServices() and ignoring any further duplicates.

My understanding is that the protobuf API requires fully-qualified service names to be unique. We don't require the fully-qualified name in ServiceDescriptor, but the current Java codegen sets ServiceDescriptor.name to "Java package + ServiceName" - this should be equivalent to the proto fully qualified name for reflection to work, as it starts with the list of services and queries for the protobufs defining those service names.

InternalHandlerRegistry.Builder keys services by their name, so implicitly makes them unique. What happens in a (mutable) HandlerRegistry is much less clear. As long as the ServiceDescriptor from a mutable HandlerRegistry has an attached ProtoFileDescriptorSupplier, the reflection service will pick it up. But how the registry actually routes requests to the underlying protobuf is undefined, so I don't think there's much the reflection service can do with duplicate service names or other discrepancies - they "should" be impossible through the normal codegen process and how the protobuf reflection API is designed to work (with fully qualified names unique).

carl-mastrangelo commented:
You should gather concrete proof that duplicates are impossible before proceeding.  Proto2 extensions are another case in protobuf where they *should* be unique but in the wild they are not.

ericgribkoff commented:
I cleaned up the tests. I also added checks to ensure that all protobuf service/method/type/extensions declarations are unique. It is necessary for service names to be unique as well: otherwise responses to the reflection file_containing_symbol="package.service" would need to return two or more file descriptors. This is easy to do, but the semantics of it are unclear for whoever is using the reflection service.

E.g., if the server has Service.MethodA defined in A.proto in the internal registry and Service.MethodB defined in B.proto in the fallback registry, the grpc_cli tool will only see one of the proto files, even if the reflection service returns both A.proto's and B.proto's file descriptors. The current implementation of the reflection service will throw an exception in such a case, rather than returning something to the client that is unusable. If we want to support these split service definitions, the changes in ProtoReflectionService are easy, but the grpc_cli tool should probably be updated as well.

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/InternalHandlerRegistry.java:
```diff
@@ -31,28 +31,44 @@
 
-  private InternalHandlerRegistry(Map<String, ServerMethodDefinition<?, ?>> methods) {
+  private InternalHandlerRegistry(List<ServerServiceDefinition> services,
+      Map<String, ServerMethodDefinition<?, ?>> methods) {
```
break line at highest semantic level.     In this case, keep all arguments on the same line.

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/AbstractServerImplBuilder.java:
```diff
@@ -127,7 +127,8 @@ public final T addService(BindableService bindableService) {
   @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
-  public final <S extends NotifyOnServerBuild & BindableService> T addService(S bindableService) {
+  public final <S extends InternalNotifyOnServerBuild & BindableService> T
+      addService(S bindableService) {
```
line break at first argument

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/InternalHandlerRegistry.java:
```diff
@@ -83,7 +86,9 @@ InternalHandlerRegistry build() {
       }
-      return new InternalHandlerRegistry(services.values(), Collections.unmodifiableMap(map));
+      return new InternalHandlerRegistry(
+          Collections.unmodifiableList(new ArrayList<ServerServiceDefinition>(services.values())),
```
you might consider making this ordered, since it makes unit tests that verify the results easier.

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -145,21 +168,25 @@ private synchronized void initFileDescriptorMaps() {
       if (serviceDescriptor.getMarshallerDescriptor() != null
-          && serviceDescriptor.getMarshallerDescriptor()
-          instanceof ProtoFileDescriptorSupplier) {
+          && serviceDescriptor.getMarshallerDescriptor() instanceof ProtoFileDescriptorSupplier) {
```
instanceof already checks for null

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -145,21 +168,25 @@ private synchronized void initFileDescriptorMaps() {
+        serviceNames.add(serviceName);
+        if (!seenFiles.contains(fileDescriptor.getName())) {
+          seenFiles.add(fileDescriptor.getName());
+          fileDescriptorsToProcess.offer(fileDescriptor);
```
check the return value?

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,352 @@
+    int extensionNumber = extension.getNumber();
+    if (!fileDescriptorsByExtensionAndNumber.containsKey(extensionName)) {
+      fileDescriptorsByExtensionAndNumber.put(extensionName,
+          new HashMap<Integer, FileDescriptor>());
```
keep args on the same line

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,352 @@
+
+  private void processType(Descriptor type, FileDescriptor fd) {
+    String typeName = type.getFullName();
+    Preconditions.checkState(!fileDescriptorsBySymbol.containsKey(typeName),
```
Would this fit on the same line if you statically imported Preconditions?

ericgribkoff replied:
It did not fit (and now the indentation is further increased).

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,352 @@
+  private synchronized void initFileDescriptorMaps() {
+    if (mapsInitialized) {
+      // Check if the server's set of services has changed
+      List<ServerServiceDefinition> currentServices = server.getServices();
```
I'm not sure you need this complexity.  @ejona86 Does this need to support dynamic Reflection service changes?

ericgribkoff replied:
In earlier discussions with @ejona86 we talked about supporting dynamic services. I dropped this check and just re-process all file descriptors for each query, which still supports dynamic services but removes some of the complexity of initialization.

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,352 @@
+            ((ProtoFileDescriptorSupplier) serviceDescriptor.getMarshallerDescriptor())
+                .getFileDescriptor();
+        String serviceName = serviceDescriptor.getName();
+        Preconditions.checkState(!serviceNames.contains(serviceName), "Service already defined: %s",
```
Same comment about line length.  I think this would fit with static imports.

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,339 @@
+
+      @Override
+      public void onNext(ServerReflectionRequest request) {
+        initFileDescriptorMaps();
```
This access is still racy.  Two requests coming in at the same time from different threads are going to be trampling each other.

ericgribkoff replied:
So a StreamObserver cannot maintain internal state without synchronization? This is done in a lot of the example code, e.g., https://github.com/grpc/grpc-java/blob/master/examples/src/main/java/io/grpc/examples/routeguide/RouteGuideServer.java#L187.

I dug into the code a little more, and it looks like the executor that is calling onNext() on a given StreamObserver will always be a serializing executor (https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ServerImpl.java#L378). My understanding is that a SerializingExecutor will avoid any race conditions in updating a StreamObservers internal state. Can this be relied upon?

I couldn't find any documentation on this besides in the Javadoc for StreamObserver, which says: "Since individual StreamObservers are not thread-safe, if multiple threads will be writing to a StreamObserver concurrently, the application must synchronize calls." I found this a little unclear - do StreamObservers have to worry about gRPC writing to a StreamObserver concurrently (in which case, why is there an "if" in the Javadoc), or is this only a worry if the application is doing something concurrently in/with a StreamObserver?

ejona86 replied:
You don't need the synchronization. I think Carl didn't realize those data structures were request-scoped. The "if" is because you don't need to "synchronize" when on a single thread, since any processing on a single thread is already coherent.

Is there a reason this initialization isn't done in the constructor of the anonymous StreamObserver? (Constructors for anonymous classes are just a block in class definition, similar to static blocks, but without the static keyword.) Also, it looks like it will "initialize" multiple times since onNext can be called multiple times in a bidi streaming RPC.

ericgribkoff replied:
It can use a constructor in the anonymous class, but I was initializing (processing) the maps on each call in case the dynamic registry services have changed. Rather than figure out how to poll for updates, it made sense just to reprocess the list of services on each call.

carl-mastrangelo replied:
@ejona86 is right.  This is not racy.  However, it is still rather inefficient.  The majority of the time the services supported will never change.  Mutable Handlers are rare, and when they are present they may not even be reflectable (such as a generic proxy). 

It's probably worth breaking this up into static data that is build at server construction time, and a lighter weight "mutable" add on that is generated per request.

ericgribkoff replied:
@carl-mastrangelo Breaking it into immutable and mutable data is fine, but that would require some method to differentiate between internal and dynamic services. The getServices() method on Server could return the internal services, and Server.getDynamicServices() (a new method) could return the dynamic ones?

@ejona86 had earlier mentioned that optimization of the reflection service wasn't particularly important at this point. Is it worthwhile differentiating between immutable and dynamic services in order to avoid this inefficiency in building the lookup tables?

ericgribkoff replied:
In the interest of making progress on this, can optimizing this experimental service be deferred to a subsequent PR? I'm happy to open an issue and assign it to myself. 

It's probably worth further consideration about the best way of differentiating dynamic and non-dynamic services.

ejona86 commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,339 @@
+   */
+  @Override
+  public void notifyOnBuild(Server server) {
+    if (this.server == null) {
```
Go ahead and assert that `notifyOnBuild` hasn't been called earlier. That would help detect bugs.

```java
Preconditions.checkState(this.server == null);
this.server = Preconditions.checkNotNull(server, "server");
```

ejona86 commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,339 @@
+  }
+
+  @Override
+  public StreamObserver<ServerReflectionRequest> serverReflectionInfo(
```
I don't see any flow control taking place. We'll need to stop reading when writing is delayed. You can cast `responseObserver` to a `ServerCallStreamObserver`, then `disableAutoInboundFlowControl()`, and then use `setOnReadyHandler()` and `isReady()` and `request()`.

ericgribkoff replied:
Is there an example anywhere that uses both `disableAutoInboundFlowControl()` and `setOnReadyHandler()`? I could not find one. How I wanted to do this was:

```
final ServerCallStreamObserver<ServerReflectionResponse> serverCallObserver =
    (ServerCallStreamObserver<ServerReflectionResponse>) responseObserver;

serverCallObserver.disableAutoInboundFlowControl();
serverCallObserver.request(1);

serverCallObserver.setOnReadyHandler(new Runnable() {
  @Override
  public void run() {
    while (serverCallObserver.isReady()) {
      serverCallObserver.request(1);
    }
  }
});
```

This doesn't work - the onReadyHandler runs in the same thread as the requestObserver, so the while loop just runs forever and nothing happens. What does work is removing the `while` loop and just requesting 1 message every time the stream observer becomes ready. This then requires the `onNext()` method in the `StreamObserver<ServerReflectionRequest>` to also check if `serverCallObserver` is ready and, if so, request another message.

I feel like this is not the intended use of the flow control API. What would be nice is to actually do the work inside of the `onReadyHandler`, e.g.:

```
serverCallObserver.setOnReadyHandler(new Runnable() {
  @Override
  public void run() {
    while (serverCallObserver.isReady()) {
      serverCallObserver.onNext(<some reflection response>);
      serverCallObserver.request(1);
    }
  }
});
```

This is similar to @ejona86's comment from Sept 9th on https://github.com/grpc/grpc-java/issues/2247, except there the message passed to `onNext()` did not depend on anything coming in from the client. Essentially, I haven't been able to figure out how to access the `ReqT request` that we get [here](https://github.com/grpc/grpc-java/blob/master/stub/src/main/java/io/grpc/stub/ServerCalls.java#L206) within the `onReadyHandler`, except by the hack described above.

Commit https://github.com/grpc/grpc-java/commit/7902017b07d09ccff6d118965b9de804d37960b0 introduced the disableAutoInboundFlowControl() and references `TestServiceImpl.streamingOutputCallManualFlowControl` as an example of the usage, but I can't find that method anywhere in Github.

ericgribkoff replied:
@ejona86 fyi

ejona86 replied:
There are no real users of the methods on ServerCallStreamObserver, but many more for ClientCall/ServerCall. You can `git grep \\.request` to see some.

Both of those snippets didn't seem quite right though. You should `request(1)` initially, and after you are done with the previous request. That will mean the code will be processing either 0 or 1 messages at any given time.

Just something thrown together quickly:
```java
final ServerCallStreamObserver<ServerReflectionResponse> serverCallObserver =
    (ServerCallStreamObserver<ServerReflectionResponse>) responseObserver;
serverCallObserver.disableAutoInboundFlowControl();
serverCallObserver.request(1);
class MyStreamObserver implements Runnable {
  ServerReflectionRequest req;
  @Override
  public void onNext(ServerReflectionRequest r) {
    assert req == null;
    req = r;
    doWork();
  }

  @Override // as OnReadyHandler
  public void run() {
    if (req != null) {
      doWork();
    }
  }

  private void doWork() {
    while (serverCallObserver.isReady()) { // This could be a do-while loop
      ServerReflectionResponse resp = computeNextResponse();
      if (resp == null) { // I'm assuming this is the signal for "done"
        // We processed the current message, so now we request another message
        req = null;
        serverCallObserver.request(1);
        return;
      }
      serverCallObserver.onNext(resp);
      // This could also be a fine place to check if done with the current message,
      // although then the loop works slightly better as while instead of do-while
      // (not enough to be bothered by, but more from a conceptual standpoint)
    }
  }
};
MyStreamObserver mso = new MyStreamObserver();
serverCallObserver.setOnReadyHandler(mso);
return mso;
```

      
      serverCallObserver.request(1);

ejona86 commented:
This PR is failing to build cleanly. It seems that's because we [only ignore cast for compileJava](https://github.com/grpc/grpc-java/blob/master/build.gradle#L134). We should change that to be similar to [what we do elsewhere](https://github.com/grpc/grpc-java/blob/master/build.gradle#L32).

ericgribkoff commented on core/src/main/java/io/grpc/HandlerRegistry.java:
```diff
@@ -42,6 +44,16 @@
+   * @throws UnsupportedOperationException if not supported by implementation
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+  public List<ServerServiceDefinition> getServices() {
```
Carl asked if I could change it from a `Collection` to a `List` or `Set`, partially to make things easier to test. But being able to return map.values() might qualify as a compelling reason to stick with `Collection`. Happy to go with whichever.

carl-mastrangelo replied:
List is the right one.  Looking further down, this is going to have allocate a list anyways from the MutableHandlerRegistry  map.

ericgribkoff commented on core/src/main/java/io/grpc/internal/InternalHandlerRegistry.java:
```diff
@@ -31,31 +31,49 @@
     // Store per-service first, to make sure services are added/replaced atomically.
-    private final HashMap<String, ServerServiceDefinition> services =
-        new HashMap<String, ServerServiceDefinition>();
+    private final SortedMap<String, ServerServiceDefinition> services =
```
The suggestion was a sorted key set would make it easier to write unit tests.

carl-mastrangelo replied:
Optional, but I would have picked LinkedHashMap.  Requiring that they be sorted is asking a lot.  Just being consistent is good enough.

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/AbstractServerImplBuilder.java:
```diff
@@ -116,6 +126,14 @@ public final T addService(BindableService bindableService) {
   }
 
```
There should be a ServerServiceDefinition overload here too right?

ericgribkoff replied:
ServerServiceDefinition is a final class. We need to check for instances of the InternalNotifyOnServerBuild interface. Can an instance of a final class also implement an interface?

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/ServerImpl.java:
```diff
@@ -176,6 +178,15 @@ public int getPort() {
+  @Override
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+  public List<ServerServiceDefinition> getServices() {
+    List<ServerServiceDefinition> services =
```
This needs to be split into static and dynamic parts.   If the fallbackRegistry is empty, there is no need to do an extra allocation.  If the mutable handler registry is not empty, you can pre allocate the correct size arraylist ahead of time.

carl-mastrangelo commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,339 @@
+  @Override
+  public StreamObserver<ServerReflectionRequest> serverReflectionInfo(
+      final StreamObserver<ServerReflectionResponse> responseObserver) {
+    return new StreamObserver<ServerReflectionRequest>() {
```
This StreamObserver is complicated enough that it probably deserves to be in its own, named class.  It can be an inner class.

carl-mastrangelo commented:
Optimize later is fine.  There are still a number of outstanding comments on this PR.

ericgribkoff commented:
The latest PR should address all of the outstanding comments, except for flow control. 

@ejona86 Does this PR need to included the server-side flow control, or can that be added later separately? 

There were a couple things I wasn't sure about with the build configuration:

*    I changed build.gradle to ignore the protoc compiler warnings for test. I don't have much experience with Gradle so the changes might not be the ideal configuration settings.
*    There's still the following warning due to Javadoc during build: "core/src/main/java/io/grpc/ServiceDescriptor.java:92: warning - Tag @link: reference not found: io.grpc.protobuf.ProtoFileDescriptorSupplier". It seems like a good idea to keep the @link in the Javadoc, but then we get the error. Should the error be suppressed or the @link removed?

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/ServerImpl.java:
```diff
@@ -176,6 +177,20 @@ public int getPort() {
+  public List<ServerServiceDefinition> getServices() {
+    if (fallbackRegistry.getServices().isEmpty()) {
+      return registry.getServices();
+    } else {
```
no need for an else here.

carl-mastrangelo commented:
One minor comment, but LGTM.  

please rebase to master, and change the commit message to match the "package: description of change" format.

ejona86 commented on core/src/main/java/io/grpc/HandlerRegistry.java:
```diff
@@ -42,6 +44,16 @@
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+  public List<ServerServiceDefinition> getServices() {
+    throw new UnsupportedOperationException("Not implemented");
```
This exception is not being handled in `ServerImpl.getServices()`. Maybe it should just return `Collections.emptyList()`? May want to do the same with `Server`.

ejona86 commented on core/src/main/java/io/grpc/util/MutableHandlerRegistry.java:
```diff
@@ -64,6 +67,12 @@ public boolean removeService(ServerServiceDefinition service) {
+  @Override
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+  public List<ServerServiceDefinition> getServices() {
+    return Collections.unmodifiableList(new ArrayList<ServerServiceDefinition>(services.values()));
```
Put a comment here that this doesn't necessarily return a consistent view of the map. That's fine because it is "safe" and no ConcurrentModificationException is thrown by ConcurrentHashMap, but it's subtle so it's probably good to call it out in one way or another. If it were a different map implementation, you would probably have needed to use `services.values().toArray()`.

ejona86 commented on core/src/main/java/io/grpc/internal/ServerImpl.java:
```diff
@@ -179,12 +178,17 @@ public int getPort() {
-        new ArrayList<ServerServiceDefinition>(registry.getServices());
-    services.addAll(fallbackRegistry.getServices());
-    return Collections.unmodifiableList(services);
+    if (fallbackRegistry.getServices().isEmpty()) {
```
Save the services returned here to avoid calling `fallbackRegistry.getServices()` three times, which isn't free.

ejona86 commented on services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java:
```diff
@@ -0,0 +1,345 @@
+
+    @Override
+    public void onNext(ServerReflectionRequest request) {
+      initFileDescriptorMaps();
```
I still don't get why this isn't loaded in the constructor. We can deal with it later in a followup PR.

ejona86 commented on core/src/main/java/io/grpc/ServerBuilder.java:
```diff
@@ -86,17 +86,6 @@
   public abstract T addService(BindableService bindableService);
 
   /**
-   * Adds a NotifyOnServerBuild service implementation to the handler registry. When build() is
```
Add some documentation to `addService` mentioning `NotifyOnServerBuild`.

ericgribkoff commented:
retest this please
